### PR TITLE
Fixing bugs

### PR DESCRIPTION
**Pull request by vasthecat**

### CODE REVIEW

**Comment by vasthecat** [comment]
Многопоточный режим иногда может зависать при завершении работы. Возможно, что-то не так с мьютексом tasks_mutex в функции mt_worker.

File: main.c [hunk 1]
```diff
@@ -143,20 +147,29 @@ bruteforce_rec(struct task_t *task,
...
+}
+
 bool
 bruteforce_iter(struct task_t *task,
                 struct config_t *config,
                 void *context,
                 password_handler_t handler)
 {
-    size_t size = strlen(config->alphabet) - 1;
+    size_t size = task->to - task->from;
```
Comment (hnarkaytis) [inline main.c]: WTF?

File: main.c [hunk 2]
```diff
@@ -190,15 +203,19 @@ singlethreaded(struct task_t *task, struct config_t *config)
     struct st_context_t context;
     context.hash = config->hash;
     context.cd.initialized = 0;
-    bool found = false;
+    context.config = config;
+
+    task->from = 0;
+    task->to = strlen(config->alphabet) - 1;
```
Comment (hnarkaytis) [inline main.c]: 'to' is a position in password template. It has nothing to do with length of the alphabet

File: main.c [hunk 3]
```diff
@@ -209,16 +226,32 @@ void *
...
-    data.initialized = 0;
     while (true)
     {
         struct task_t task;
         queue_pop(&context->queue, &task);
 
-        char *hashed = crypt_r(task.password, context->hash, &data);
-        if (strcmp(hashed, context->hash) == 0)
+        task.from = task.to;
+        task.to = strlen(config->alphabet) - 1;
```
Comment (hnarkaytis) [inline main.c]: The same problem as in init

File: main.c [hunk 4]
```diff
@@ -266,13 +300,16 @@ multithreaded(struct task_t *task, struct config_t *config)
         pthread_create(&threads[i], NULL, mt_worker, (void *) &context);
     }
 
+    task->from = 0;
+    task->to = strlen(config->alphabet) - 1 - 2;
```
Comment (hnarkaytis) [inline main.c]: +1

File: main.c [hunk 5]
```diff
@@ -111,6 +112,7 @@ struct st_context_t
 {
     char *hash;
     struct crypt_data cd;
+    struct config_t *config;
```
Comment (hnarkaytis) [inline main.c]: Doesn't look to be required here

File: sem.h [hunk 6]
```diff
@@ -17,12 +18,21 @@ sem_init(sem_t *sem, int pshared, int value)
...
+{
+    sem_t *sem = (sem_t *) arg;
+    pthread_mutex_unlock(&sem->value_mutex);
+}
+
 void
 sem_wait(sem_t *sem)
 {
     pthread_mutex_lock(&sem->value_mutex);
+    pthread_cleanup_push(&exit_func, (void *) sem);
```
Comment (hnarkaytis) [inline sem.h]: Could be reduced to 
`pthread_cleanup_push ((void (*) (void*))pthread_mutex_unlock, &sem->value_mutex);`

File: sem.h [hunk 7]
```diff
@@ -17,12 +18,21 @@ sem_init(sem_t *sem, int pshared, int value)
...
     while (sem->value == 0)
         pthread_cond_wait(&sem->sem_cond, &sem->value_mutex);
+    pthread_cleanup_pop(1);
```
Comment (hnarkaytis) [inline sem.h]: pthread_cleanup_pop (!0);

File: main.c (hunk 1 above)
Comment (vasthecat) [reply main.c]: I already fixed it (and 3 issues below) in fourth commit (9804552).

Comment (vasthecat) [reply main.c]: I already fixed it (and 3 issues below) in fourth commit (9804552).

File: main.c [hunk 8]
```diff
@@ -274,14 +300,16 @@ multithreaded(struct task_t *task, struct config_t *config)
         pthread_create(&threads[i], NULL, mt_worker, (void *) &context);
     }
 
-    bool found = false;
+    task->from = 0;
```
Comment (hnarkaytis) [inline main.c]: It's easier to iterate suffix in producer and prefix in consumer.
in producer:
task->from = 2
task->to = config->length;
in consumer:
task->to = task->from;
task->from = 0;

Common practice is to treat upper bound as non-inclusive. I.e.
for (i = task->from; i < task->to; ++i)

**Review by hnarkaytis — APPROVED** [review]

**Comment by vasthecat** [comment]
Works correctly all the time on linux. When compiled on linux with semaphores from sem.h and crypt_r from crypt-macos still works correctly.

Sometimes halts on macOS. Can't consistently reproduce, but can be seen with `for i in {1..1000}; do ./brute -m; done`.

File: main.c [hunk 9]
```diff
@@ -169,21 +169,21 @@ bruteforce_iter(struct task_t *task,
                 void *context,
                 password_handler_t handler)
 {
-    size_t size = task->to - task->from;
+    size_t size = strlen(config->alphabet) - 1;
     int a[config->length];
```
Comment (hnarkaytis) [inline main.c]: task->to is a more correct upper limit

**Review by hnarkaytis — APPROVED** [review]

File: main.c (hunk 9 above)
Comment (vasthecat) [reply main.c]: Seems right

**Pull request merged**